pkg/server/registry/apigroups/acorn/images/copy.go

```diff
@@ -0,0 +1,353 @@
+			data, err := json.Marshal(update)
+			if err != nil {
+				panic("failed to marshal update: " + err.Error())
+			}
+			if err := conn.WriteMessage(websocket.TextMessage, data); err != nil {
+				logrus.Errorf("Error writing pull status: %v", err)
+				break
+			}
+		}
+
+		_ = conn.CloseHandler()(websocket.CloseNormalClosure, "")
+	}), nil
+}
+
+func (i *ImageCopy) ConnectMethods() []string {
+	return []string{"GET"}
+}
+
+func (i *ImageCopy) Validate(ctx context.Context, args apiv1.ImageCopy, namespace string) error {
+	if err := imagesystem.IsNotInternalRepo(ctx, i.client, namespace, args.Source); err != nil {
+		return err
+	}
+
+	if err := imagesystem.IsNotInternalRepo(ctx, i.client, namespace, args.Dest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *ImageCopy) ImageCopy(ctx context.Context, namespace, sourceImage, destImage string, srcAuth, dstAuth *apiv1.RegistryAuth, force bool) (<-chan ImageProgress, error) {
+	sourceRef, err := i.getImageReference(ctx, sourceImage, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if sourceRef.Context().RegistryStr() == images.NoDefaultRegistry {
+		return nil, fmt.Errorf("missing registry name (i.e. docker.io, ghcr.io) from source %s", sourceImage)
+	}
+
+	destRef, err := name.ParseReference(destImage, name.WithDefaultRegistry(images.NoDefaultRegistry))
+	if err != nil {
+		return nil, err
+	}
+
+	if sourceRef.Context().RegistryStr() == images.NoDefaultRegistry {
+		return nil, fmt.Errorf("missing registry name (i.e. docker.io, ghcr.io) from destination %s", destImage)
+	}
+
+	sourceOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, sourceRef.Context(), srcAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	destOpts, err := images.GetAuthenticationRemoteOptionsWithLocalAuth(ctx, destRef.Context(), dstAuth, i.client, namespace, i.transportOpt)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceIndex, err := remote.Index(sourceRef, sourceOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if !force {
+		if err := errIfImageExistsAndIsDifferent(sourceIndex, destRef, destOpts); err != nil {
+			return nil, err
+		}
+	}
+
+	progress := make(chan ggcrv1.Update)
+	// progress gets closed by remote.WriteIndex, so this second channel is so that
+	// we can control closing the result channel in case we need to write an error.
+	progress2 := make(chan ImageProgress)
+	destOpts = append(destOpts, remote.WithProgress(progress))
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		for update := range progress {
+			var errString = ""
+			if update.Error != nil {
+				errString = update.Error.Error()
+			}
+			progress2 <- ImageProgress{
+				Total:       update.Total,
+				Complete:    update.Complete,
+				Error:       errString,
+				CurrentTask: fmt.Sprintf("Copying %s to %s", sourceImage, destImage),
+			}
+		}
+	}()
+
+	go func() {
+		defer func() {
+			wg.Wait()
+			close(progress2)
+		}()
```

Comment:
nit: this (and the `WaitGroup`) are not necessary. You can close the `progress2` channel at the end of the above goroutine.